service/history/api/workflow_id_dedup.go

```diff
@@ -100,6 +104,39 @@ func ResolveDuplicateWorkflowID(
 	return nil, nil
 }
 
+func resolveJustStartedDuplicatedWorkflows(
+	shardContext shard.Context,
+	currentWorkflowStartTime time.Time,
+	newRunID string,
+	workflowID string,
+) (UpdateWorkflowActionFunc, error) {
+	// we are going to use "grace period"
+	// to prevent multiple calls to start worklfows with the same ID
+	// if new workflow is starting earlier then that period - we will not terminate old worklfow,
+	// but rather didn't start the new one
+
+	gracePeriod := shardContext.GetConfig().WorkflowIdReuseMinimalInterval()
+	now := shardContext.GetTimeSource().Now().UTC()
+	timeSinceStart := now.Sub(currentWorkflowStartTime.UTC())
+
+	if gracePeriod == 0 || gracePeriod < timeSinceStart {
+		return terminateWorkflowAction(newRunID)
+	}
+
+	// there is a grace period, and old worklfow start time is within that period
+	// do not start new worklfow
```

Comment:
How about?

```suggestion
	// Since there is a grace period, and the current workflow's start time is within that period,
	// abort the entire request.
```